lib/autoupdate/agent/process.go

```diff
@@ -0,0 +1,175 @@
+	switch {
+	case code < 0:
+		return trace.Errorf("unable to determine if systemd service is active")
+	case code > 0:
+		s.Log.WarnContext(ctx, "Teleport systemd service not running.")
+		return trace.Wrap(ErrNotNeeded)
+	}
+	// Attempt graceful reload of running service.
+	code = s.systemctl(ctx, slog.LevelError, "reload", s.ServiceName)
+	switch {
+	case code < 0:
+		return trace.Errorf("unable to attempt reload of Teleport systemd service")
+	case code > 0:
+		// Graceful reload fails, try hard restart.
+		code = s.systemctl(ctx, slog.LevelError, "try-restart", s.ServiceName)
+		if code != 0 {
+			return trace.Errorf("hard restart of Teleport systemd service failed")
+		}
+		s.Log.WarnContext(ctx, "Teleport ungracefully restarted. Connections potentially dropped.")
+	default:
+		s.Log.InfoContext(ctx, "Teleport gracefully reloaded.")
+	}
+
+	// TODO(sclevine): Ensure restart was successful and verify healthcheck.
+
+	return nil
+}
+
+// Sync systemd service configuration by running systemctl daemon-reload.
+// See Process interface for more details.
+func (s SystemdService) Sync(ctx context.Context) error {
+	if err := s.checkSystem(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+	code := s.systemctl(ctx, slog.LevelError, "daemon-reload")
+	if code != 0 {
+		return trace.Errorf("unable to reload systemd configuration")
+	}
+	return nil
+}
+
+// checkSystem returns an error if the system is not compatible with this process manager.
+func (s SystemdService) checkSystem(ctx context.Context) error {
+	_, err := os.Stat("/run/systemd/system")
+	if errors.Is(err, os.ErrNotExist) {
+		s.Log.ErrorContext(ctx, "This system does not support systemd, which is required by the updater.")
+		return trace.Wrap(ErrNotSupported)
+	}
+	return trace.Wrap(err)
+}
+
+// systemctl returns a systemctl subcommand, converting the output to logs.
+// Output sent to stdout is logged at debug level.
+// Output sent to stderr is logged at the level specified by errLevel.
+func (s SystemdService) systemctl(ctx context.Context, errLevel slog.Level, args ...string) int {
+	cmd := exec.CommandContext(ctx, "systemctl", args...)
+	stderr := &lineLogger{ctx: ctx, log: s.Log, level: errLevel}
+	stdout := &lineLogger{ctx: ctx, log: s.Log, level: slog.LevelDebug}
+	cmd.Stderr = stderr
+	cmd.Stdout = stdout
+	err := cmd.Run()
+	stderr.Flush()
+	stdout.Flush()
+	code := cmd.ProcessState.ExitCode()
+	if code == 255 {
+		code = -1
+	}
+	if err != nil {
+		s.Log.Log(ctx, errLevel, "Failed to run systemctl.",
+			"args", args,
+			"code", code,
+			"error", err)
+	}
+	return code
+}
+
+// lineLogger logs each line written to it.
+type lineLogger struct {
+	ctx   context.Context
+	log   *slog.Logger
+	level slog.Level
+
+	last bytes.Buffer
+}
+
+func (w *lineLogger) Write(p []byte) (n int, err error) {
+	lines := bytes.Split(p, []byte("\n"))
+	// Finish writing line
+	if len(lines) > 0 {
+		n, err = w.last.Write(lines[0])
```

Comment:
`n` returned is the amount of byte written to the log. If the last line was incomplete, `n` will count the last write lenght plus the lenght of `p` before the first newline.

If I do this:
```
lineLogger.write([]byte("foo"))
lineLogger.write([]byte("bar"))
lineLogger.write([]byte("baz\n"))
```

The last call will return n=10 because len("foobarbaz\n"), however `len(p) == 4`. According to the Writer interface godoc:

> Write writes len(p) bytes from p to the underlying data stream. It returns the number of bytes written from p (0 <= n <= len(p)) and any error encountered that caused the write to stop early. Write must return a non-nil error if it returns n < len(p). Write must not modify the slice data, even temporarily.

We would return 0, 0, then 10, which is not compliant with the Writer interface. I think the proper behaviour is to always return `len(p)`.

Reply:
I found an error here (Line 153 should be `n += 1`), but in the case that the `w.last.Write`s fail, `n` still needs to return the number of accepted bytes, which may be `< len(p)`.

The code actually returns 3, 3, 13, because it's counting all accepted bytes (just accidentally double-counting the initial bytes). With line 153 adjusted, it works: https://go.dev/play/p/09Xpi4XCZpI

Will add unit tests 🙂

Reply:
> The code actually returns 3, 3, 13

The last write does not seem compatible with this part of the doc:

> the number of bytes written from p (0 <= n <= len(p))

Reply:
I meant the code before I made the change. If you click the link, fixing line 153 returns 3, 3, 4.